executor/src/witgen/jit/symbolic_expression.rs

```diff
@@ -0,0 +1,319 @@
+
+use crate::witgen::range_constraints::RangeConstraint;
+
+/// A value that is known at run-time, defined through a complex expression
+/// involving known cells or variables and compile-time constants.
+/// Each of the sub-expressions can have its own range constraint.
+#[derive(Debug, Clone)]
+pub enum SymbolicExpression<T: FieldElement, V> {
+    /// A concrete constant value known at compile time.
+    Concrete(T),
+    /// A symbolic value known at run-time, referencing either a cell or a local variable.
+    Variable(V, Option<RangeConstraint<T>>),
+    BinaryOperation(
+        Rc<Self>,
+        BinaryOperator,
+        Rc<Self>,
+        Option<RangeConstraint<T>>,
+    ),
+    UnaryOperation(UnaryOperator, Rc<Self>, Option<RangeConstraint<T>>),
+}
+
+#[derive(Debug, Clone)]
+pub enum BinaryOperator {
+    Add,
+    Sub,
+    Mul,
+    /// Finite field division.
+    Div,
+    /// Integer division, i.e. convert field elements to unsigned integer and divide.
+    IntegerDiv,
+    BitAnd,
+    BitOr,
+}
+
+#[derive(Debug, Clone)]
+pub enum UnaryOperator {
+    Neg,
+}
+
+impl<T: FieldElement, V> SymbolicExpression<T, V> {
+    pub fn from_var(name: V) -> Self {
+        SymbolicExpression::Variable(name, None)
+    }
+
+    pub fn is_known_zero(&self) -> bool {
+        self.try_to_number().map_or(false, |n| n.is_zero())
+    }
+
+    pub fn is_known_one(&self) -> bool {
+        self.try_to_number().map_or(false, |n| n.is_one())
+    }
+
+    pub fn is_known_minus_one(&self) -> bool {
+        self.try_to_number().map_or(false, |n| n == -T::from(1))
+    }
+
+    pub fn range_constraint(&self) -> Option<RangeConstraint<T>> {
+        match self {
+            SymbolicExpression::Concrete(v) => Some(RangeConstraint::from_value(*v)),
+            SymbolicExpression::Variable(.., rc)
+            | SymbolicExpression::BinaryOperation(.., rc)
+            | SymbolicExpression::UnaryOperation(.., rc) => rc.clone(),
+        }
+    }
+
+    pub fn try_to_number(&self) -> Option<T> {
+        match self {
+            SymbolicExpression::Concrete(n) => Some(*n),
+            SymbolicExpression::Variable(..)
+            | SymbolicExpression::BinaryOperation(..)
+            | SymbolicExpression::UnaryOperation(..) => None,
+        }
+    }
+}
+
+impl<T: FieldElement, V: Display> Display for SymbolicExpression<T, V> {
+    fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
+        match self {
+            SymbolicExpression::Concrete(n) => {
+                if n.is_in_lower_half() {
+                    write!(f, "{n}")
+                } else {
+                    write!(f, "-{}", -*n)
+                }
+            }
+            SymbolicExpression::Variable(name, _) => write!(f, "{name}"),
+            SymbolicExpression::BinaryOperation(lhs, op, rhs, _) => {
+                write!(f, "({lhs} {op} {rhs})")
+            }
+            SymbolicExpression::UnaryOperation(op, expr, _) => write!(f, "{op}({expr})"),
+        }
+    }
+}
```

Comment:
I think the use of parenthesis is inconsistent here: For `BinaryOperation`, you put them around the result, but not the operands, for `UnaryOperation`, you do the opposite, and for negative concrete numbers you don't put any.

So you'd get strings like this:
`-(((a * -5) + -(b)))`

I think it makes sense to always put parenthesis around the result? Then this would be:
`(-((a * (-5)) + (-b)))`

Reply:
BTW, are these implementations for informational purposes, or are these supposed to be valid Rust?

Reply:
No, just informational, I'll add a comment.

Reply:
I'll put parentheses, but not around unary, since it's still unambiguous